### DISCUSSION TITLE
[SDK] TWAIN device doesn't show in SDK. But does show in NAPS2 app

### DISCUSSION THREAD

**Discussion by outomator** [cyanfish/naps2#370] [Development]
Hello, 

I copied the TWAIN example from [https://github.com/cyanfish/naps2/blob/master/NAPS2.Sdk.Samples/TwainSample.cs](url) but my device isn't listed using `await controller.GetDeviceList(Driver.Twain)`. However, I am able to find and use the device in the NAPS2 app using TWAIN drivers.

Side note:  Using WIA, `GetDeviceList()` the scanner shows. But performing a scan causes an error. NAPS2 app throws this error below upon selection.

System.InvalidCastException: Specified cast is not valid.
   at IWiaItem2 NAPS2.Wia.Native.IWiaDevMgr2.SelectDeviceDlg(IntPtr hwndParent, int lDeviceType, int lFlags, ref string pbstrDeviceID)
   at uint NAPS2.Wia.Native.NativeWiaMethods.SelectDevice2(IntPtr deviceManagerPtr, IntPtr hwnd, int deviceType, int flags, out string deviceId, out IntPtr device)
   at WiaDevice NAPS2.Wia.WiaDeviceManager.PromptForDevice(IntPtr parentWindowHandle)
   at T NAPS2.EtoForms.EtoInvoker.InvokeGet<T>(Func<T> func)+() => { }
   at T NAPS2.EtoForms.EtoInvoker.InvokeGet<T>(Func<T> func)
   at async Task<ScanDevice> NAPS2.Scan.ScanPerformer.PromptForDevice(ScanOptions options) x 2

Thanks!

**Comment by cyanfish** [comment]
What SDK version are you using?

**Reply by outomator** [reply]
Latest 1.0.1

**Comment by cyanfish** [comment]
Specifically, do you mean the list returned from `await controller.GetDeviceList(Driver.Twain)` is empty? If everything is the latest version it's supposed have the same behaviour between the app and SDK.

What version of Windows and .NET? In the NAPS2 app, in your profile settings, under Advanced, do you have anything other than "Default" selected under Twain Implementation? Is the SDK running as a normal user application, not a service or anything like that?

**Reply by outomator** [reply]
Correct, the list is empty.

Tried on a Windows 10 and Windows 11 machine. Used both .NET8 and .NET6. And yea everything is set to default in advanced.

**Reply by outomator** [reply]
And forgot to mention I'm running this through Visual Studio

**Reply by cyanfish** [reply]
In the NAPS2 app, if you select "Old DSM" under "Twain Implementation" does it show up?

**Reply by outomator** [reply]
Yes, scan works fine using "Old DSM". Same as before

**Reply by outomator** [reply]
Edit: Tried again, and when setting it to "Old DSM", no scanning devices are found.

Any idea what else may be causing this? Or are there any other troubleshooting steps I could try? Thanks!

**Comment by outomator** [comment]
Would love to get this working if you have any other suggestions. Thanks!

**Reply by shallowain1** [reply]
Try to change the build options of your project to "x86" (right click on your project | properties | build | general | platform target).
